File: design_analysis/benchmarks/etl/100m/single-thread/pandas/task_4_aggregate_region.py
#!/usr/bin/env python3
"""Task 4 (Pandas): sum(price * quantity). Matches 100m/arche/task_4 checksum semantics."""

import sys
import time
import pandas as pd

DEFAULT_CSV = "design_analysis/benchmarks/etl/data/data_100m.csv"


def main(csv_path):
    start = time.perf_counter()
    df = pd.read_csv(csv_path, usecols=["price", "quantity"])
    df["revenue"] = df["price"] * df["quantity"]
    checksum = df["revenue"].sum()
    elapsed = time.perf_counter() - start
    print(f"task4_checksum: {checksum}")
    print(f"task4_time: {elapsed}")


if __name__ == "__main__":
    csv_path = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_CSV
    main(csv_path)

File: design_analysis/benchmarks/etl/100m/single-thread/c/task_4_aggregate_region.c
/* Task 4 (C baseline): sum(price * quantity).
 * Same compute as Task 1; named "aggregate_region" for symmetry with arche/pandas/etc.
 */

#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

#include "fast_parse.h"

#define DEFAULT_CSV "design_analysis/benchmarks/etl/data/data_100m.csv"

static double now_sec(void) {
	struct timespec ts;
	clock_gettime(CLOCK_MONOTONIC, &ts);
	return ts.tv_sec + ts.tv_nsec * 1e-9;
}

int main(int argc, char **argv) {
	const char *path = argc > 1 ? argv[1] : DEFAULT_CSV;

	int fd = open(path, O_RDONLY);
	if (fd < 0) {
		fprintf(stderr, "open(%s): %s\n", path, strerror(errno));
		return 1;
	}
	struct stat sb;
	if (fstat(fd, &sb) < 0) {
		close(fd);
		return 1;
	}
	size_t size = (size_t)sb.st_size;
	char *base = mmap(NULL, size, PROT_READ, MAP_PRIVATE, fd, 0);
	if (base == MAP_FAILED) {
		close(fd);
		return 1;
	}

	double t0 = now_sec();

	char *p = (char *)memchr(base, '\n', size);
	if (!p) {
		munmap(base, size);
		close(fd);
		return 1;
	}
	p++;
	char *end = base + size;

	double revenue_sum = 0.0;
	const char *q = p;
	while (q < end) {
		while (q < end && *q != ',')
			q++; /* skip timestamp */
		q++;
		double price = fast_atof(&q);
		q++;
		long quantity = fast_atol(&q);
		revenue_sum += price * (double)quantity;
		while (q < end && *q != '\n')
			q++;
		q++;
	}

	double elapsed = now_sec() - t0;

	munmap(base, size);
	close(fd);

	printf("task4_checksum: %.6f\n", revenue_sum);
	printf("task4_time: %.6f\n", elapsed);
	return 0;
}
